File: leetcode/DFS/leetcode-200/200.go
package dfs

func numIslands(grid [][]byte) int {
	if grid == nil || len(grid) == 0 {
		return 0
	}

	row, col := len(grid), len(grid[0])

	flag := make([][]bool, row)
	for i := 0; i < row; i++ {
		flag[i] = make([]bool, col)
	}
	count := 0
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if grid[i][j] == '1' && !flag[i][j] {
				count++
				dfs(grid, flag, i, j)
			}
		}
	}
	return count
}

func dfs(grid [][]byte, flag [][]bool, row, col int) {
	if row < 0 || col < 0 || row >= len(grid) || col >= len(grid[0]) {
		return
	}

	if flag[row][col] {
		return
	}

	if grid[row][col] == '0' {
		return
	}

	c := [4][2]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	flag[row][col] = true
	for i := 0; i < 4; i++ {
		dfs(grid, flag, row+c[i][0], col+c[i][1])
	}
}

File: leetcode/DFS/leetcode-200/200.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int row = grid.length;
        int col = grid[0].length;
        boolean[][] flag = new boolean[row][col];
        int count = 0;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (grid[i][j] == '1' && !flag[i][j]) {
                    count++;
                    dfs(grid, flag, i, j);
                }
            }
        }
        return count;
    }

    private void dfs(char[][] grid, boolean[][] flag, int row, int col) {
        if (row < 0 || col < 0 || row >= grid.length || col >= grid[0].length) {
            return;
        }

        if (flag[row][col]) {
            return;
        }
        if (grid[row][col] == '0') {
            return;
        }
        flag[row][col] = true;
        int[][] c = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
        for (int i = 0; i < 4; i++) {
            dfs(grid, flag, row + c[i][0], col + c[i][1]);
        }
    }
}